man/nkillall.8
.TH nkillall 8 "Dec 28, 2009"
.SH NAME
nkillall \- send signals to all processes;  write messages to users
.SH SYNOPSIS
.B nkillall
-[vq] [-s secs] [-M|W mesg] [-signum] [-E prog [arg[s]]
.SH DESCRIPTION
.B nkillall
sends signals
to all processes skipping it's own PID.  
It is a simple program that does not use \fI/proc\fR FS.
It uses kill(2) and execve(2) syscalls.  It is designed as
a replacement of the SysVinit killall5(8) command.

.SH USAGE
It is not allowed to use
.B \-h
(signal SIGHUP) as first option!  If you really want this then start:
.B nkillall
\-s0 \-hup ...

.TP
.B \-v 
verbose mode;  \-vv means more verbose
.TP
.B \-q
quiet mode;  ignores SIGINT signal;  Ctrl-C does't work!
.TP
.B \-s 
\fIsecs\fR 
.br
sleep \fIsecs\fR
.TP
.B \-M 
\fImesg\fR
.br
write \fImesg\fR to stdout.  The message can include escape
symbols: 
\\NNN, \\a, \\b, \\e, \\c, \\f, \\n, \\r, \\t, \\v, \\\\
like in C language.  Last message can be written again with
\-M% or \-W%.  Example:

.B nkillall
\-M'\\n\\aPlease logout!  The system is going down...' \\
         \-W% \-s2 -W% \-s4 \-W% \-s12 \\
         \-M'\\tSending TERM signal' \-term ...
.TP
.B \-W
\fImesg\fR
.br
write \fImesg\fR to all logged users.  The syntax is the same as \-M.
.B nkillall
determines logged users by looking at
.I /var/run/utmp.

.br
In my opinion 
.I /var/run/utmp
is unreliable!
Other programs like who(1), wall(1), login(1) also use this file.
If you feel that yours is insecure, better don't use the \-W option.

.TP
.B \-signum
send all processes the singal \fIsignum\fR.  
Some important signals can be encodded.  For example SIGTERM can be typed with
\-15 or \-term.
Only the first letter is important!
The command
.B kill \-l
lists the signal numbers.
.TP
.B \-E
.I /path/to/program [arg[s]]
.br
execve(2) the program with some arguments.
This option must be last!  The program will have the same PID as
.B nkillall.

.SH NOTE
The options are applied immediately.  Their position is important.
The next two examples work differently:

nkillall \-q -s2 \-M'sending signal SIGTERM' \-15 ...
.br
nkillall \-s2 \-M'sending signal SIGTERM' \-15 \-q ...

Use the options \-v[v] \-q after 
.B nkillall
immediately.

.SH WARNING
.B nkillall
also kills the script in which it is included.
Use it in scripts like:

#!/bin/sh
.br
before commands
.br
exec nkillall [options] \-E/path/to/prog prog_opt(s)

An example of wrong usage is the following:

#!/bin/sh
.br
before commands
.br
nkillall ...
.br
after commands

.SH EXAMPLES
.B nkillall
\-M'Sending all processes SIGTERM ...' \-s2 \-15 \\
.br
         \-M'Sending all processes SIGKILL ...' \-s6 \-9 \\
.br
         \-s1 \-E /path/to/program arg1 arg2 ...

.B nkillall 
\-v \-s1 \-15 \-cont \-s6 \-kill \-s1 \-E/path/to/prog arg(s) ...

.B nkillall 
\-vv \-pause \-s59 \-continue \-M'Hello world!'

.SH FILES
.I /var/run/utmp

.SH AUTHOR
.B nkillall
is included in ninit package and can be downloaded from
.br
.I http://riemann.frmi.uni-sofia.bg/ninit/

.SH "SEE ALSO"
ninit(8), kill(1), kill(2), execve(2), utmp(5) 
.br
init(8), killall5(8)


CHANGES
0.14:   2010-01-19 00:27:01
  added nkillall.8  
  gid is not default anymore.  see ninitfeatures.h
  uid supports 22:33:100:200:300
  nkillall: \e escape
  shutdown: set cron off
  nkillall: signals with letters like nsvc
  shutdown and nkillall: makes more sync(2), flag -q
  added THANKS ;-)
  added helpers: remove procfs;  thanks to them snvc is smaller
  nsvc options -onumber -unumber; option -E
  Cron has option -C
  /etc/ninit/.nsvc_help
  nsvc, shutdown, reload uses variables NINIT_MEMORY and NINIT_HOME
  nsvc: option -S
  shutdown: option -s
  manpages for all programs in /sbin ;-)
  update manpages (Boris Grozev)
  Makefile: man_txt
  serdo.c: read only one file again
  better sleep in open_inout.h
  serdo.c improvments; buildin: killall5, simple echo
  nsvc.c: -C+number
  scripts/conf
  ninit.8: NONROOT USAGE, using /sbin/init with ninit
  ninit-scan: (Wayne Marshall)

0.13.7:   2009-12-16 10:47:50
  bootlog.c (flush_root)
  0.13.1.4 renamed to 0.13.2
  service start 'end' (if it exixts; X_OK) after finishing
  fixed a bug in shutdown.c (thx to Jan)
  With flag -s ninit-shutdown don't start halt and reboot services
  shutdown.c:  improvements; flags -S -q
  shutdown.c:  removed flags -s -q; new flags -E -v -T
  added nkillall
  shutdown.c:  options -m and -E work together
  nkillall print escape like echo

0.13.1:   2009-06-27 12:59:38
  updated bootlog
  Makefile logs 'make tests' in tests_log using boolog
  fixing quotes in serdo

0.13:   2009-01-10 18:11:42
  it's possible to set parameters in wait.  example: some_service:180:3
  install-bin print help and string errors
  better sync mode.  maxsync is removed.  echo 200 > sync
  updated scripts/ninit_test.sh
  comments (#) in depends, params, environ, wait, softlimit
  install-bin (verbose mode)
  nsvc -V
  errmsg_put.c; buffer_*.c
  setup, rsetup, sys-rsetup starts with args: $1=service $2=service_pid
  EXTRACT_* trick (automatic create header files)
  stuct process is 20 bytes on x86_64 (see struct_root.h)
  changed BIFFER_INIT macro
  utmp_io.c
  Makefile: install_other
  serdo is intslled only with: make install_other
  t_write.h uses PIPE_BUF (thx Laurent Bercot)
  nsvc.8 improvements
  ninit_test.sh creates services in ./etc/ninit/
  ninit-huge
  added printf.c and ninit.spec
  make FLAG_DEBUG=no
  removed flag X_OK in some syscalls access.
  better addprocess

0.12.1:   2008-01-03 17:15:08
  ninit-depends.  convert directory to file (depends.dir -> depends)
  removed unused stat.h in headers
  updated serdo (static)
  two environ vars: NINIT_HOME, INIT_HOME.  (Thanks to Stamatis Mitrofanis)
  ninit logo: /etc/ninit/.sync
  install-bin uses chown32 instead chown on some systems
  err.c and err_b.c uses the macros va_start and va_arg
  fu.c (u &= 0xffffffff)

0.12:
  almost all arrays are moved in stack.  small data and bss sections.
  assembler functions (i386) for some DJB functions.
  modifications in Makefile.
  100 Euro security guarantee.

0.11.2:
  ninit-reload accept agument -e (change environ) -E file_env
  program ~/sys/update
  service flags: pause, pause-wait
  service flag cron:  a:b[:c]
    can contain many lines.
  flag -K for nsvc.
  pidfilehack --> ninit-pidfile.
  make i386 -- build static daemons (don't use dietlibc)
  Makefile is smaller and simpler
  gcc-4.1.2 works now.
  install-bin
  stuct mem in initialized in the main program.

0.11.1:
  ninit-reload accepts args (-r3 -r5 -r33 ...)
  ninit-inittab sets the variables NINIT_RUNLEVEL, INIT_VERSION
  if /etc/minit/{in|out} FIFO's exist then make symbolic links.
  new run flags: sys-rsetup, pidfile
  if the service sysvinit exist, ninit create /dev/initctl 
  ninit-reload and nsvc checks for environ NINIT_HOME
  nsvc -[drR] ALL - change all services.
    Change service ALL with: nsvc -[drR] ops ALL
  new program: ninit-sysvinit.  configuration file: sysvinit-timeout
    fimeout[:fork-mode].  example:  echo 600:0 > sysvinit-timeout
  ninit catch signal SIGPWR
  new run flags: alarm, maxsync
  nsvc output is different
  shell script ninit-service
  new program run-wait

0.11:
  Sun Jun 17 22:08:22 EEST 2007
  I decided to rename all without changes.
  The name ninit must be spelled nano-init.

0.9.11:
  rewritten by Nikola Vladov

0.9.1:
  fix embarassing typo in msvc (Gelu G. Lupas)

0.9:
  See minit CHANGES - http://www.fefe.de/minit


nkillall.c
#include <unistd.h>
#include <fcntl.h>
#include <time.h>
#include <signal.h>
#include <alloca.h>
#include <utmp.h>
#include <errno.h>
#include "ninitfeatures.h"
#include "process_defs.h"
#include "utmp_defs.h"
#include "uid.h"
#include "check_opt.h"

static int cfg_verbose;

void sighandler(int sig) { (void)sig; }

static void deep_sleep(char *x) {
  time_t t = x_atoi(x);
  time_t deedline = time(0) + t;
  if (t<=0) return;
  if (cfg_verbose>1) msg("Sleeping: ", x);
  while (1) {
    nano_sleep(1,0);
    if (time(0) >= deedline) break;
  }
}

static void print_escape(char *x, int wall) {
  char ch, *y, *start, *from = "abefnrtv\\c", *to= "\a\b\033\f\n\r\t\v\\";
  unsigned long len;
  start = alloca(str_len(x) + 3);
  for (y=start; (ch=*x); x++) {
    if (ch == '\\') {
      unsigned long u;
      len = scan_8ulong(x+1, &u);
      if (len) { /* 012...9 */
	ch = u;
	x += len;
      } else {
	len = str_chr(from, x[1]);
	switch (from[len]) {
	  case 0: break;
	  case 'c': goto do_it;
	  default: /* \a\b... */ ++x; ch = to[len];
	}
      }
    }
    *y++ = ch;
  }
  *y++ = '\n';

 do_it:
  len = y-start;

  if (wall==0) write(1, start, len);
  else {
    struct utmp u;
    int fd, r;
    char line[8 + sizeof(u.ut_line)];

    fd = open(_PATH_UTMP, O_RDONLY | O_NOCTTY);
    if (fd < 0) return;

    while (utmp_io(fd, &u, F_RDLCK)) {
      if (u.ut_type != USER_PROCESS) continue;
      if (u.ut_user[0] == 0 || u.ut_pid < 2 ||
	  u.ut_line[0] == 0 || u.ut_line[0] == '/') continue; 

      y = line;
      y += str_copyn(y, "/dev/", 8);
      y += str_copyn(y, u.ut_line, sizeof(u.ut_line));
      y[0] = 0;
      if (line[str_chr(line, '.')]) continue;

      if (kill(u.ut_pid, 0) && errno == ESRCH) continue;
      r = open(line, O_WRONLY | O_NOCTTY | O_NDELAY);
      if (r < 0) continue;

      write(r, start, len);
      close(r);
    }
    close(fd);
  }
}

#define Kill_Help \
"Usage:\n",me, \
" -[vq] [-s secs] [-M|W mesg] [-signum] ... [-E program [arg[s]]\n" \
"\t-s secs:\tsleep secs\n" \
"\t-M mesg:\twrite mesg to stdout\n" \
"\t-W mesg:\twrite mesg to logged users\n" \
"\t-sugnum:\tsignal number\n" \
"\t-v[v]:\t\tbe verbose\n" \
"\t-q:\t\tquiet mode; ignores SIGINT signal\n\n" \
"Some signals can be codded. Only the first letter is important!\n" \
"\t-p:\tSIGSTOP\n" \
"\t-c:\tSIGCONT\n" \
"\t-h:\tSIGHUP\n" \
"\t-a:\tSIGALRM\n" \
"\t-i:\tSIGINT\n" \
"\t-t:\tSIGTERM\n" \
"\t-k:\tSIGKILL\n\n" \
"Option -E must be last.  Examples:\n  ", \
me," -s2 -M'Sending all processes SIGTERM ...' -term -cont \\\n" \
"\t-s6 -M'Sending all processes SIGKILL ...' -9 \\\n" \
"\t-s1 -E /path/to/program arg1 arg2 ...\n\n  ", \
me," -q -vv -s1 -15 -cont -s6 -kill -s1" \
" -E/path/to/prog arg(s) ..."

int main(int argc, char **argv) {
  char *x, *me = argv[0];
  char *last_msg = "\\c";
  errmsg_iam(me);
  if (argc<2) {
  help:
    errmsg_iam(0);
    carp(Kill_Help);
    _exit(1);
  }
  argv++;
  x = argv[0];
  if (x[0]=='-' && x[1]=='h') goto help; /* -h... --help */

  opendevconsole();

  /* ignore next signals */
  set_sa(SIGQUIT); set_sa(SIGCHLD); set_sa(SIGHUP );
  set_sa(SIGTSTP); set_sa(SIGTTIN); set_sa(SIGTTOU);

  while (argv[0] && argv[0][0]=='-') {
    int sig=0, cfg_wall=0;
    char *y = argv[0] + 1;
    if ((unsigned int)(*y - '0') < 10) { 
      sig = x_atoi(y);
      goto again;
    }
    
    switch(*y) {
    case 'v': cfg_verbose++; if (y[1]=='v') cfg_verbose++; break;
    case 'q': set_sa(SIGINT); break;
    case 's': chk_opt(argv,x); deep_sleep(x); break;
    case 'W': ++cfg_wall;
    case 'M': chk_opt(argv,x);
      if (x[0] == '%' && x[1] == 0) x = last_msg;
      else last_msg = x;
      print_escape(x, cfg_wall); 
      break;
    case 'E':
      chk_opt(argv,x); *argv = x;
      execve(argv[0], argv, environ);
      carp("Unable to exec: ", argv[0]);
      _exit(1);
    default:
#ifdef NSVC_SIGNAL_CODED
      {
	unsigned char *S, *Sig = (unsigned char *)NSVC_SIGNAL_CODED;
	for (S=Sig; *S; S += 2)
	  if ((unsigned char)*y == *S) { sig = S[1]; goto again; }
      }
#else
#define kk(C,S) case C: sig=S; break
	kk('t',SIGTERM); kk('a',SIGALRM); kk('p',SIGSTOP); kk('c',SIGCONT);
	kk('h',SIGHUP);  kk('i',SIGINT);  kk('k',SIGKILL);
#endif
      goto help;
    }

  again:
    if (sig > 0) {
      sync();
      if (cfg_verbose) msg("Sending signal: ", y);
      if (kill(-1, sig)) carp("Unable to send signal: ", y);
    }
    argv++;
  }
  sync();
  return 0;
}
